README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# LWFBrook90R

Run the
[LWF-BROOK90](https://www.lwf.bayern.de/boden-klima/wasserhaushalt/index.php)
hydrological model in R.

## Installation

It is recommended to use the latest **stable version**, which can be
found at <https://github.com/pschmidtwalter/LWFBrook90R/releases>. If
`LWFBrook90R_*.tar.gz` / `LWFBrook90R_*.zip` are available for the
latest release, you can download it and install it on Linux / Windows in
the usual way.

Otherwise (or if you want to install from source) you can install a
tagged version directly from the Github repo. On Windows this requires
[Rtools](https://cran.r-project.org/bin/windows/Rtools/). The tag of the
latest release (e.g. `v0.3.4`) can also be found on the
[releases-page](https://github.com/pschmidtwalter/LWFBrook90R/releases).

``` r
if(!require("remotes")) install.packages("remotes")
remotes::install_github(repo="pschmidtwalter/LWFBrook90R", build_vignettes=TRUE,
                        ref="v0.3.4") 
```

<br />

The current **development version** can be installed in the same way

``` r
if(!require("remotes")) install.packages("remotes")
remotes::install_github(repo="pschmidtwalter/LWFBrook90R", build_vignettes=TRUE) 
```

After installation, take a look at the vignette with
`vignette("intro_lwfbrook90r")`.

## Basic usage

Load LWFBrook90R:

``` r
library(LWFBrook90R)
```

To run an example, load meteorological and soil data from the Solling
Beech Experimental site distributed with the package:

``` r
data("slb1_meteo")
data("slb1_soil")
```

Set up lists containing default model control options and model
parameters:

``` r
options.b90 <- setoptions_LWFB90()
param.b90 <- setparam_LWFB90()
```

Set start and end dates in model control options:

``` r
options.b90$startdate <- as.Date("2002-01-01")
options.b90$enddate <- as.Date("2003-12-31")
```

Derive soil hydraulic properties from soil physical properties using a
pedotransfer function:

``` r
soil <- cbind(slb1_soil, hydpar_puh2(clay = slb1_soil$clay,
                                     silt = slb1_soil$silt,
                                     sand = slb1_soil$sand,
                                     bd = slb1_soil$bd,
                                     oc.pct = slb1_soil$c_org))
```

Run LWF-Brook90 with the created model input objects and capture results
in `b90.results.slb1`:

``` r
b90.results.slb1 <- runLWFB90(project.dir = "example_run_b90/",
                              param.b90 = param.b90,
                              options.b90 = options.b90,
                              climate = slb1_meteo,
                              soil = soil)
str(b90.results.slb1, max.level = 1)
```

## Status

### R-Code

The package works as intended and is fully documented. However, there
are some points to be accomplish in the near future:

  - [x] enable use of Clapp-Hornberger hydraulic parameterization in
    addition to the default Mualem-van Genuchten
  - [x] Use of sub-day resolution precipitation interval data.
  - [ ] implement unit tests (functionality is currently tested through
    vignette and examples)
  - [ ] Run the `check` with Travis.

### Fortran-Code

  - [x] Model output results tested against the output from the original
    ‘b90.exe’ Windows command line Fortran program.
  - [x] Cleaning up `declared but not used` variables
  - [x] Use of sub-day resolution precipitation interval data.
  - [x] Optional writing of model log-file that contains the former
    commandline-feed.
  - [x] Implementation of an error-routine. All STOP/EXIT commands
    removed from Fortran code.
  - [x] Return model output directly back to R, without writing text
    files.

## Authors

Paul Schmidt-Walter, Volodymyr Trotsiuk, Klaus Hammel, Martin Kennel,
Tony Federer.

Tony Federer’s original [Brook90 Fortran 77
code](http://www.ecoshift.net/brook/b90doc.html) (Brook90\_v3.1F,
License: CC0) was enhanced by Klaus Hammel and Martin Kennel at Bavarian
State Institute of Forestry (LWF) around the year 2000. Since then,
LWF-BROOK90 is distributed by
[LWF](https://www.lwf.bayern.de/boden-klima/wasserhaushalt/index.php)
upon request as a pre-compiled Fortran command line program together
with an MS Access User Interface. In 2019, Volodymyr Trotsiuk converted
the Fortran 77 code to Fortran 95 and implemented the connection to R.
Paul Schmidt-Walter’s *brook90r* package for LWF-Brook90 input data
generation, model execution and result processing was adapted and
extended to control this interface function.

## License

GPL-3 for all Fortran and R code. brook90r has GPL-3, while LWF-Brook90
was without license until recently. Lothar Zimmermann and Stephan Raspe
(LWF), as well as all Fortran contributors agreed to assign GPL-3 to the
Fortran code.


NEWS
LWFBrook90R v0.?.? (Release date: ???)
==============
Changes:
* switched `msiterunLWFB90()` and `mrunLWFB90()` from superseded packages `snow` and `doSNOW` to `future`, `doFuture` and `progressr` for parallel computation and progress reporting thereof. Pacifies a check note and is more future-proof.


LWFBrook90R v0.3.4 (Release date: 2020-08-28)
==============

Changes
* Adjusted 'table'-method in MakeRelRootDens: The root depth distribution provided in a table is now redistributed to the soil nodes under preservation of the total root mass.
* Input changed for soilnodes-argument of MakeRelRootDens. See ?MakeRelRootDens.

Bug fixes:
* zero division error causing infinite relawat values in MISCDAY.ASC-output.
* 'betamodel'-method of MakeRelRootDens returned the increment of the cumulative root proportion at the soil nodes, instead of the relative root density, which actually is the former value divided by the layer thickness.


LWFBrook90R v0.3.0 (Release date: 2020-04-21)
==============

Changes
* new function msiterunLWFB90()
* runLWFB90(): model input (param.b90, option.b90, standprop_daily) is appended
	to the return value BEFORE evaluating the output_fun-argument. In this way,
	on-the-fly post-processing of model results including model-input is possible now.
* minor bugs fixed


LWFBrook90R v0.2.0 (Release date: 2019-06-18)
==============

Changes:
* minor bugs fixed
* 'output_fun' argument replaces 'gof_fun'-argument in runLWFB90() for more flexible output.


LWFBrook90R v0.1.0 (Release date: 2019-05-13)
==============
 first beta release: most functionality was taken from brook90r. However, functions
 and arguments were renamed, and a lot of functionality was added.


src/PFILE.h
    !*************************************************************************************
!     PFILE.INC reads parameters from the input


    ! output specifications -------
    NDAYS = INT( param( 1 ) )
    HEAT = INT( param( 2 ) )
    OP(:,:) = output(:,:)

    ! Meteorologic parameters -------
    ESLOPE = param( 3 )
    ASPECT = param( 4 )

    ! Convert to radians
    ESLOPE = ESLOPE / 57.296
    ASPECT = ASPECT / 57.296

    ALB = param( 5 )
    ALBSN = param( 6 )
    C1 = param( 7 )
    C2 = param( 8 )
    C3 = param( 9 )
    WNDRAT = param( 10 )
    FETCH = param( 11 )
    Z0W = param( 12 )
    ZW = param( 13 )

    ! Canopy parameters -------
    LWIDTH = param( 14 )
    Z0G = param( 15 )
    Z0S = param( 16 )
    LPC = param( 17 )
    CS = param( 18 )
    CZS = param( 19 )
    CZR = param( 20 )
    HS = param( 21 )
    HR = param( 22 )
    ZMINH = param( 23 )
    RHOTP = param( 24 )
    NN = param( 25 )

    ! Interception parameters -------
    RSTEMP = param( 26 )
    DURATN = pdur( 1, : )
    INTR = param( 27 )
    INTS = param( 28 )
    FRINTL = param( 29 )
    FSINTL = param( 30 )
    FRINTS = param( 31 )
    FSINTS = param( 32 )
    CINTRL = param( 33 )
    CINTRS = param( 34 )
    CINTSL = param( 35 )
    CINTSS = param( 36 )

    ! Snow parameters -------
    MELFAC = param( 37 )
    CCFAC = param( 38 )
    LAIMLT = param( 39 )
    SAIMLT = param( 40 )
    GRDMLT = param( 41 )
    MAXLQF = param( 42 )
    KSNVP = param( 43 )
    SNODEN = param( 44 )

    ! leaf parameters affecting PE -------
    GLMAX = param( 45 )
    CR = param( 46 )
    GLMIN = param( 47 )
    RM = param( 48 )
    R5 = param( 49 )
    CVPD = param( 50 )
    TL = param( 51 )
    T1 = param( 52 )
    T2 = param( 53 )
    TH = param( 54 )

    ! plant parameters affecting soil-water supply -------
    MXKPL = param( 55 )
    MXRTLN = param( 56 )
    inirlen = param( 57 )
    inirdep = param( 58 )
    rgrorate = param( 59 )
    rgroper = param( 60 )
    FXYLEM = param( 61 )
    PSICR = param( 62 )
    RTRAD = param( 63 )
    NOOUTF = INT( param( 64 ) )

    if( FXYLEM .GE. 0.990d0 ) FXYLEM = 0.990d0
    if( inirlen .LT. 0.010d0 ) inirlen=0.010d0
    if( inirdep .LT. 0.010d0 ) inirdep=0.010d0

    ! soil parameters -------
    NLAYER = INT( param( 65 ) )
    nmat = INT( param( 66 ) )
    ILAYER = INT( param( 67 ) )
    QLAYER = INT( param( 68 ) )
    IMODEL = INT( param( 69 ) )

    if ( (NLAYER .GT. ML) .OR. (ILAYER .GT. NLAYER) .OR. (QLAYER .GT. NLAYER)) then
        write(10,*) 'Failure of QLAYER and ILAYER .LE. NLAYER .LE. ML'
        go to 999
    end if

    DO 201 I=1, nmat
        if (imodel .eq. 0) then
            ParMat(1,I) = soil_materials(I,2)
            ParMat(2,I) = soil_materials(I,3)
            ParMat(4,I) = soil_materials(I,4)
            ParMat(9,I) = soil_materials(I,5)
            ParMat(3,I) = soil_materials(I,6)
            ParMat(10,I) = soil_materials(I,7)
            StonefMat(I) = soil_materials(I,8)
            ParMat(5,I) = 0.0d0
            ParMat(6,I) = 0.0d0
            ParMat(7,I) = 0.0d0
            ParMat(8,I) = 0.0d0
        end if

        if(imodel .eq. 1) then
            ParMat(1,I) = soil_materials(I,2)
            ParMat(10,I) = soil_materials(I,3)
            ParMat(7,I) = soil_materials(I,4)
            ParMat(8,I) = soil_materials(I,5)
            ParMat(6,I) = soil_materials(I,6)
            ParMat(9,I) = soil_materials(I,7)
            StonefMat(I) = soil_materials(I,8)
            ParMat(2,I) = 0.0d0
            ParMat(4,I) = 0.0d0
            ParMat(5,I) = 0.0d0
            !..it's a sin......
            ! Hard default [mm d-1] for saturated hydraulic conductivity at field capacity
            ParMat(3,I) = 2
        end if
201 CONTINUE


    DO 202 I=1, NLAYER
        if ( HEAT .EQ. 1 ) then
            dep(I) = soil_nodes(I,2)
            THICK(I) = soil_nodes(I,3)
            mat(I) = INT( soil_nodes(I,4) )
            PSIM(I) = soil_nodes(I,5)
            frelden(I) = soil_nodes(I,6)
            ! TemperatureNew(i) = soil_nodes(I,7) we don't have it in the input file!!!

            if(I .LT. NLAYER) then
                MUE(I) = THICK(I) / ( THICK(I) + THICK(I+1) )
                ZL(I) = 0.50d0 * ( THICK(I) + THICK(I+1) )
            else
                MUE(I) = 0.50d0
                ZL(I) = THICK(I)
            end if

            TMean(I) = 0.0d0
        else
            dep(I) = soil_nodes(I,2)
            THICK(I) = soil_nodes(I,3)
            mat(I) = INT( soil_nodes(I,4) )
            PSIM(I) = soil_nodes(I,5)
            frelden(I) = soil_nodes(I,6)
        end if
202 CONTINUE

    depmax = dep(1) - THICK(1) / 1000.0d0
    !..from material-specific to layer-specific parameter values
    DO 203 I=1, NLAYER
        DO 204 J=1, MPAR
            PAR(J,I) = ParMat(J,mat(I))
204     CONTINUE
        STONEF(I) = StonefMat( mat(I) )
203 CONTINUE

    RSSA = param( 70 )
    RSSB = param( 71 )

    ! find thickness of maximum root zone
    ifoundfirst = 0
    i1=0
    i2=0
    DO 205 I=1, NLAYER
        if(ifoundfirst .EQ. 0) then
            if( frelden(I) .GE. 1.e-6) then
                ifoundfirst = 1
                i1 = I
            end if
        end if
        if( frelden(I) .GE. 1.e-6) i2 = I-1
205 CONTINUE

    if( (ifoundfirst .eq. 1) .and. (i2 .lt. i1) ) i2 = NLAYER
    !write(*,*)' i1= ',i1,' i2= ',i2

    DO 206 I=1,NLAYER
        tini(I) = 1.e+20
        if( (I .GE. i1) .AND. (I .LE. i2) ) frelden(I) = MAX( frelden(I), 1.01e-6)
        if( (frelden(I) .GE. 1.e-6) .AND. (depmax-dep(I)) .LE. inirdep) then
            tini(I) = 0.0d0
        end if
        if( (frelden(I) .GE. 1.e-6) .AND. (depmax-dep(I)) .GT. inirdep) then
            if(rgrorate .GT. 0) then
                tini(I) = (depmax-dep(I)-inirdep)/rgrorate
            end if
        end if
        !write(*,*)'dep= ',dep(I),' tini= ',tini(I)
206 CONTINUE

    ! flow parameters -------
    INFEXP = param( 72 )
    BYPAR = INT( param( 73 ) )
    QFPAR = param( 74 )
    QFFC = param( 75 )
    IMPERV = param( 76 )
    DSLOPE = param( 77 )
    LENGTH = param( 78 )
    DRAIN = param( 79 )
    GSC = param( 80 )
    GSP = param( 81 )

    ! integration parameters -------
    DTIMAX = param( 82 )
    DSWMAX = param( 83 )
    DPSIMX = param( 84 )

    ! heat flow -------
    ! we assign some so compilation does not complain
    TopInfT = 1
!       if (HEAT .EQ. 1) then
!        READ (12,*) Comment
!        READ (12,*) tTop, Comment
!        tTop=tTop
!        READ (12,*) tBot, Comment
!        tBot=tBot
!        READ (12,*) TopInfT, Comment
!        READ (12,*) BotInfT, Comment
!        READ (12,*) kTopT, Comment
!        READ (12,*) kBotT, Comment
!        DO 207 I = 1, 7
!         READ (12,*) Comment
! 207    CONTINUE
!        DO 208 I = 1, nmat
!         READ (12,*) ilay, SV(I), OV(I), HB1(I), HB2(I), HB3(I)
!         TPar(1,I) = SV(I)
!         TPar(2,I) = OV(I)
!         TPar(3,I) = THDis
! C        thermal conductivities -- transfer from [J m-1 s-1 K-1] to  [J mm-1 d-1 K-1]
!         TPar(4,I) = HB1(I) * 86.400
!         TPar(5,I) = HB2(I) * 86.400
!         TPar(6,I) = HB3(I) * 86.400
! C         volumetric heat capacities for solid, water and organic -- transfer from [MJ m-2 mm-1 K-1] to [J mm-3 K-1]
!         TPar(7,I) = CVSOL
!         TPar(8,I) = CVORG
!         TPar(9,I) = CVLQ
! 208    CONTINUE
!        READ (12,*) C
!       end if
